src/eth1spec/spec.py

```diff
@@ -0,0 +1,354 @@
+
+    Returns
+    -------
+    verified : `bool`
+        True if the header is correct, False otherwise.
+    """
+    raise NotImplementedError()  # TODO
+
+
+def apply_body(
+    state: State,
+    coinbase: Address,
+    block_number: Uint,
+    block_gas_limit: Uint,
+    block_time: U256,
+    block_difficulty: Uint,
+    transactions: List[Transaction],
+    ommers: List[Header],
+) -> Tuple[Uint, Root, State]:
+    """
+    Executes a block.
+
+    Parameters
+    ----------
+    state : `eth1spec.eth_types.State`
+        Current account state.
+    coinbase : `eth1spec.eth_types.Address`
+        Address of account which receives block reward and transaction fees.
+    block_number : `eth1spec.base_types.Uint`
+        Position of the block within the chain.
+    block_gas_limit : `eth1spec.base_types.Uint`
+        Initial amount of gas available for execution in this block.
+    block_time : `eth1spec.base_types.U256`
+        Time the block was produced, measured in seconds since the epoch.
+    block_difficulty : `eth1spec.base_types.Uint`
+        Difficulty of the block.
+    transactions : `List[eth1spec.eth_types.Transaction]`
+        Transactions included in the block.
+    ommers : `List[eth1spec.eth_types.Header]`
+        Headers of ancestor blocks which are not direct parents (formerly
+        uncles.)
+
+    Returns
+    -------
+    gas_available : `eth1spec.base_types.Uint`
+        Remaining gas after all transactions have been executed.
+    root : `eth1spec.eth_types.Root`
+        State root after all transactions have been executed.
+    state : `eth1spec.eth_types.State`
+        State after all transactions have been executed.
+    """
+    gas_available = block_gas_limit
+    receipts = []
+
+    if coinbase not in state:
+        state[coinbase] = EMPTY_ACCOUNT
+
+    for tx in transactions:
+        assert tx.gas <= gas_available
+        sender_address = recover_sender(tx)
+
+        env = evm.Environment(
+            caller=sender_address,
+            origin=sender_address,
+            block_hashes=[],
+            coinbase=coinbase,
+            number=block_number,
+            gas_limit=block_gas_limit,
+            gas_price=tx.gas_price,
+            time=block_time,
+            difficulty=block_difficulty,
+            state=state,
+        )
+
+        gas_used, logs = process_transaction(env, tx)
+        gas_available -= gas_used
+
+        receipts.append(
+            Receipt(
+                post_state=Root(trie.root(trie.map_keys(state))),
+                cumulative_gas_used=(block_gas_limit - gas_available),
+                bloom=b"\x00" * 256,
+                logs=[],
+            )
+        )
+
+    state[coinbase].balance += BLOCK_REWARD
+
+    receipts_map = {
+        bytes(rlp.encode(Uint(k))): v for (k, v) in enumerate(receipts)
+    }
+    receipts_y = trie.map_keys(receipts_map, secured=False)
+    return (block_gas_limit - gas_available), trie.root(receipts_y), state
```

Comment:
*nitpick*

Slightly more verbose (and thus maybe more readable) by removing the return statement inlining.

```python
...
gas_remaining = block_gas_limit - gas_available
post_root = trie.root(receipts_y)

return (gas_remaining, post_root, state)
```